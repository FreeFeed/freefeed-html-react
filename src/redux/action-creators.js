export const SERVER_ERROR = 'SERVER_ERROR'

export function serverError(error) {
  return {
    type: SERVER_ERROR,
    error
  }
}

export const UNAUTHENTICATED = 'UNAUTHENTICATED'

export function unauthenticated(payload) {
  return {
    type: UNAUTHENTICATED,
    payload: {...payload, authToken: ''},
  }
}

export function requiresAuth(action) {
  return action.apiRequest && !action.nonAuthRequest
}

export const request = (type) =>`${type}_REQUEST`
export const response = (type) => `${type}_RESPONSE`
export const fail = (type) => `${type}_FAIL`

import * as Api from '../services/api'

export const WHO_AM_I = 'WHO_AM_I'

export function whoAmI(){
  return {
    type: WHO_AM_I,
    apiRequest: Api.getWhoAmI,
  }
}

export const HOME = 'HOME'

export function home(offset = 0){
  return {
    type: HOME,
    apiRequest: Api.getHome,
    payload: {offset},
  }
}

export const DISCUSSIONS = 'DISCUSSIONS'

export function discussions(offset = 0){
  return {
    type: DISCUSSIONS,
    apiRequest: Api.getDiscussions,
    payload: {offset},
  }
}

export const DIRECT = 'DIRECT'

export function direct(offset = 0){
  return {
    type: DIRECT,
    apiRequest: Api.getDirect,
    payload: {offset},
  }
}

export const GET_USER_FEED = 'GET_USER_FEED'

export function getUserFeed(username, offset = 0){
  return {
    type: GET_USER_FEED,
    apiRequest: Api.getUserFeed,
    payload: {username, offset},
  }
}

export const SHOW_MORE_COMMENTS = 'SHOW_MORE_COMMENTS'

export function showMoreComments(postId){
  return {
    type: SHOW_MORE_COMMENTS,
    apiRequest: Api.getPostWithAllCommentsAndLikes,
    payload: {postId},
  }
}

export const SHOW_MORE_LIKES = 'SHOW_MORE_LIKES'

export function showMoreLikes(postId){
  return {
    type: SHOW_MORE_LIKES,
    payload: {postId},
  }
}

export const SHOW_MORE_LIKES_ASYNC = 'SHOW_MORE_LIKES_ASYNC'

export function showMoreLikesAsync(postId){
  return {
    type: SHOW_MORE_LIKES_ASYNC,
    apiRequest: Api.getLikesOnly,
    payload: {postId},
  }
}

export const SHOW_MORE_LIKES_SYNC = 'SHOW_MORE_LIKES_SYNC'

export function showMoreLikesSync(postId){
  return {
    type: SHOW_MORE_LIKES_SYNC,
    payload: {postId},
  }
}

export const TOGGLE_EDITING_POST = 'TOGGLE_EDITING_POST'

export function toggleEditingPost(postId, newValue){
  return {
    type: TOGGLE_EDITING_POST,
    payload: {postId, newValue},
  }
}

export const CANCEL_EDITING_POST = 'CANCEL_EDITING_POST'

export function cancelEditingPost(postId, newValue){
  return {
    type: CANCEL_EDITING_POST,
    payload: {postId, newValue},
  }
}

export const SAVE_EDITING_POST = 'SAVE_EDITING_POST'

export function saveEditingPost(postId, newPost){
  return {
    type: SAVE_EDITING_POST,
    apiRequest: Api.updatePost,
    payload: {postId, newPost},
  }
}

export const DELETE_POST = 'DELETE_POST'

export function deletePost(postId){
  return {
    type: DELETE_POST,
    apiRequest: Api.deletePost,
    payload: {postId},
  }
}

export const TOGGLE_COMMENTING = 'TOGGLE_COMMENTING'

export function toggleCommenting(postId){
  return {
    type: TOGGLE_COMMENTING,
    postId,
  }
}

export const ADD_COMMENT = 'ADD_COMMENT'

export function addComment(postId, commentText){
  return {
    type: ADD_COMMENT,
    apiRequest: Api.addComment,
    payload: {
      postId,
      commentText,
    }
  }
}

export const LIKE_POST = 'LIKE_POST'

export function likePost(postId, userId){
  return {
    type: LIKE_POST,
    apiRequest: Api.likePost,
    payload: {
      postId,
      userId
    }
  }
}

export const UNLIKE_POST = 'UNLIKE_POST'

export function unlikePost(postId, userId){
  return {
    type: UNLIKE_POST,
    apiRequest: Api.unlikePost,
    payload: {
      postId,
      userId
    }
  }
}

export const HIDE_POST = 'HIDE_POST'

export function hidePost(postId) {
  return {
    type: HIDE_POST,
    apiRequest: Api.hidePost,
    payload: {
      postId
    }
  }
}

export const UNHIDE_POST = 'UNHIDE_POST'

export function unhidePost(postId) {
  return {
    type: UNHIDE_POST,
    apiRequest: Api.unhidePost,
    payload: {
      postId
    }
  }
}

export const DISABLE_COMMENTS = 'DISABLE_COMMENTS'

export function disableComments(postId) {
  return {
    type: DISABLE_COMMENTS,
    apiRequest: Api.disableComments,
    payload: {
      postId
    }
  }
}

export const ENABLE_COMMENTS = 'ENABLE_COMMENTS'

export function enableComments(postId) {
  return {
    type: ENABLE_COMMENTS,
    apiRequest: Api.enableComments,
    payload: {
      postId
    }
  }
}

export const TOGGLE_EDITING_COMMENT = 'TOGGLE_EDITING_COMMENT'

export function toggleEditingComment(commentId){
  return {
    type: TOGGLE_EDITING_COMMENT,
    commentId,
  }
}

export const SAVE_EDITING_COMMENT = 'SAVE_EDITING_COMMENT'

export function saveEditingComment(commentId, newCommentBody){
  return {
    type: SAVE_EDITING_COMMENT,
    apiRequest: Api.updateComment,
    payload: {commentId, newCommentBody}
  }
}

export const DELETE_COMMENT = 'DELETE_COMMENT'

export function deleteComment(commentId){
  return {
    type: DELETE_COMMENT,
    payload: {commentId},
    apiRequest: Api.deleteComment,
  }
}

export const CREATE_POST = 'CREATE_POST'

export function createPost(feeds, postText, attachmentIds, more) {
  return {
    type: CREATE_POST,
    payload: {feeds, postText, attachmentIds, more},
    apiRequest: Api.createPost
  }
}

export const ADD_ATTACHMENT_RESPONSE = 'ADD_ATTACHMENT_RESPONSE'

export function addAttachmentResponse(postId, attachments) {
  return {
    type: ADD_ATTACHMENT_RESPONSE,
    payload: {postId, attachments}
  }
}

export const REMOVE_ATTACHMENT = 'REMOVE_ATTACHMENT'

export function removeAttachment(postId, attachmentId) {
  return {
    type: REMOVE_ATTACHMENT,
    payload: {postId, attachmentId}
  }
}

export const SIGN_IN_CHANGE = 'SIGN_IN_CHANGE'

export function signInChange(username, password){
  return {
    type: SIGN_IN_CHANGE,
    username,
    password,
  }
}

export const SIGN_IN = 'SIGN_IN'

export function signIn(username, password){
  return {
    type: SIGN_IN,
    apiRequest: Api.signIn,
    nonAuthRequest: true,
    payload: {
      username,
      password,
    },
  }
}

export const SIGN_IN_EMPTY = 'SIGN_IN_EMPTY'

export function signInEmpty(){
  return {
    type: SIGN_IN_EMPTY
  }
}

export const SIGN_UP_CHANGE = 'SIGN_UP_CHANGE'

export function signUpChange(signUpData){
  return {
    type: SIGN_UP_CHANGE,
    ...signUpData,
  }
}

export const SIGN_UP = 'SIGN_UP'

export function signUp(signUpData){
  return {
    type: SIGN_UP,
    apiRequest: Api.signUp,
    nonAuthRequest: true,
    payload: { ...signUpData },
  }
}

export const SIGN_UP_EMPTY = 'SIGN_UP_EMPTY'

export function signUpEmpty(errorMessage){
  return {
    type: SIGN_UP_EMPTY,
    message: errorMessage
  }
}

export const UPDATE_USER = 'UPDATE_USER'

export function updateUser(id, screenName, email, isPrivate){
  return {
    type: UPDATE_USER,
    payload: {id, screenName, email, isPrivate},
    apiRequest: Api.updateUser,
  }
}

export const USER_SETTINGS_CHANGE = 'USER_SETTINGS_CHANGE'

export function userSettingsChange(payload){
  return {
    type: USER_SETTINGS_CHANGE,
    payload,
  }
}

export const UPDATE_PASSWORD = 'UPDATE_PASSWORD'

export function updatePassword(payload){
  return {
    type: UPDATE_PASSWORD,
    apiRequest: Api.updatePassword,
    payload,
  }
}

export const UPDATE_USER_PHOTO = 'UPDATE_USER_PHOTO'

export function updateUserPhoto(picture){
  return {
    type: UPDATE_USER_PHOTO,
    apiRequest: Api.updateProfilePicture,
    payload: {picture},
  }
}

export const GET_SINGLE_POST = 'GET_SINGLE_POST'

export function getSinglePost(postId){
  return {
    type: GET_SINGLE_POST,
    apiRequest: Api.getPostWithAllCommentsAndLikes,
    payload: {postId},
  }
}

const userChangeAction = (type, apiRequest) => (payload) => {
  return {
    payload,
    type,
    apiRequest,
  }
}

export const BAN = 'BAN'

export const ban = userChangeAction(BAN, Api.ban)

export const UNBAN = 'UNBAN'

export const unban = userChangeAction(UNBAN, Api.unban)

export const SUBSCRIBE = 'SUBSCRIBE'

export const subscribe = userChangeAction(SUBSCRIBE, Api.subscribe)

export const UNSUBSCRIBE = 'UNSUBSCRIBE'

export const unsubscribe = userChangeAction(UNSUBSCRIBE, Api.unsubscribe)

export const GET_USER_COMMENTS = 'GET_USER_COMMENTS'

export function getUserComments(username, offset = 0){
  return {
    type: GET_USER_COMMENTS,
    payload:{username, offset},
    apiRequest: Api.getUserComments,
  }
}

export const GET_USER_LIKES = 'GET_USER_LIKES'

export function getUserLikes(username, offset = 0){
  return {
    type: GET_USER_LIKES,
    payload:{username, offset},
    apiRequest: Api.getUserLikes,
  }
}

export const EXPAND_SEND_TO = 'EXPAND_SEND_TO'

export function expandSendTo(){
  return {
    type: EXPAND_SEND_TO
  }
}

export const TOGGLE_HIDDEN_POSTS = 'TOGGLE_HIDDEN_POSTS'

export function toggleHiddenPosts() {
  return {
    type: TOGGLE_HIDDEN_POSTS
  }
}

export const SUBSCRIBERS = 'SUBSCRIBERS'

export function subscribers(username){
  return {
    type: SUBSCRIBERS,
    apiRequest: Api.getSubscribers,
    payload: {username},
  }
}

export const feedGeneratingActions = [HOME, DISCUSSIONS, GET_USER_FEED, GET_USER_COMMENTS, GET_USER_LIKES, DIRECT]
export const feedRequests = feedGeneratingActions.map(request)
export const feedResponses = feedGeneratingActions.map(response)
export const feedFails = feedGeneratingActions.map(fail)
export const isFeedRequest = action => feedRequests.indexOf(action.type) !== -1
export const isFeedResponse = action => feedResponses.indexOf(action.type) !== -1
export const isFeedFail = action => feedFails.indexOf(action.type) !== -1

export const userChangeActions = [SUBSCRIBE, UNSUBSCRIBE]
export const userChangeResponses = userChangeActions.map(response)
export const isUserChangeResponse = action => userChangeResponses.indexOf(action.type) !== -1
