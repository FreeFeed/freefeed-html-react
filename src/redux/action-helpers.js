import {
  HOME, DISCUSSIONS, DIRECT, GET_USER_FEED, GET_USER_COMMENTS, GET_USER_LIKES, GET_SEARCH, GET_BEST_OF, GET_EVERYTHING,
  SIGN_UP, WHO_AM_I, SUBSCRIBE, UNSUBSCRIBE, GET_SUMMARY, GET_USER_SUMMARY,
  UPDATE_USER, UPDATE_USER_PREFERENCES, MEMORIES, GET_USER_MEMORIES,
} from './action-types';


export const request = (type) => `${type}_REQUEST`;
export const response = (type) => `${type}_RESPONSE`;
export const fail = (type) => `${type}_FAIL`;

export const feedGeneratingActions = [HOME, DISCUSSIONS, GET_USER_FEED, GET_USER_COMMENTS, GET_USER_LIKES, DIRECT, GET_SEARCH, GET_BEST_OF, GET_EVERYTHING, GET_SUMMARY, GET_USER_SUMMARY, MEMORIES, GET_USER_MEMORIES];
const userFeedGeneratingActions = [GET_USER_FEED, GET_USER_COMMENTS, GET_USER_LIKES, GET_USER_SUMMARY, GET_USER_MEMORIES];

export const isFeedGeneratingAction = (action) => feedGeneratingActions.includes(action.type);
const isUserFeedGeneratingAction = (action) => userFeedGeneratingActions.includes(action.type);

export const feedRequests = feedGeneratingActions.map(request);
export const feedResponses = feedGeneratingActions.map(response);
export const feedFails = feedGeneratingActions.map(fail);
export const isFeedRequest = (action) => feedRequests.includes(action.type);
export const isFeedResponse = (action) => feedResponses.includes(action.type);
export const isFeedFail = (action) => feedFails.includes(action.type);
const isUserFeedRequest = (action) => userFeedGeneratingActions.map(request).includes(action.type);

export const userChangeActions = [SIGN_UP, WHO_AM_I, SUBSCRIBE, UNSUBSCRIBE, UPDATE_USER, UPDATE_USER_PREFERENCES];
export const userChangeResponses = userChangeActions.map(response);
export const isUserChangeResponse = (action) => userChangeResponses.includes(action.type);

export function requiresAuth(action) {
  return action.apiRequest && !action.nonAuthRequest;
}

const unRequest = (type) => type.replace(/_REQUEST$/, '');

export function getFeedName(action) {
  if (!isFeedGeneratingAction(action) && !isFeedRequest(action)) {
    return "unknown";
  }

  if (isUserFeedGeneratingAction(action) || isUserFeedRequest(action)) {
    return `${unRequest(action.type)}_${action.payload.username}`;
  }

  return unRequest(action.type);
}
