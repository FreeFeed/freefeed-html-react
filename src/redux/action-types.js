export const SERVER_ERROR = 'SERVER_ERROR'
export const UNAUTHENTICATED = 'UNAUTHENTICATED'
export const WHO_AM_I = 'WHO_AM_I'
export const HOME = 'HOME'
export const DISCUSSIONS = 'DISCUSSIONS'
export const DIRECT = 'DIRECT'
export const GET_USER_FEED = 'GET_USER_FEED'
export const SHOW_MORE_COMMENTS = 'SHOW_MORE_COMMENTS'
export const SHOW_MORE_LIKES = 'SHOW_MORE_LIKES'
export const SHOW_MORE_LIKES_ASYNC = 'SHOW_MORE_LIKES_ASYNC'
export const SHOW_MORE_LIKES_SYNC = 'SHOW_MORE_LIKES_SYNC'
export const TOGGLE_EDITING_POST = 'TOGGLE_EDITING_POST'
export const CANCEL_EDITING_POST = 'CANCEL_EDITING_POST'
export const SAVE_EDITING_POST = 'SAVE_EDITING_POST'
export const DELETE_POST = 'DELETE_POST'
export const TOGGLE_COMMENTING = 'TOGGLE_COMMENTING'
export const UPDATE_COMMENTING_TEXT = 'UPDATE_COMMENTING_TEXT'
export const ADD_COMMENT = 'ADD_COMMENT'
export const LIKE_POST = 'LIKE_POST'
export const UNLIKE_POST = 'UNLIKE_POST'
export const HIDE_POST = 'HIDE_POST'
export const UNHIDE_POST = 'UNHIDE_POST'
export const TOGGLE_MODERATING_COMMENTS = 'TOGGLE_MODERATING_COMMENTS'
export const DISABLE_COMMENTS = 'DISABLE_COMMENTS'
export const ENABLE_COMMENTS = 'ENABLE_COMMENTS'
export const TOGGLE_EDITING_COMMENT = 'TOGGLE_EDITING_COMMENT'
export const SAVE_EDITING_COMMENT = 'SAVE_EDITING_COMMENT'
export const DELETE_COMMENT = 'DELETE_COMMENT'
export const CREATE_POST = 'CREATE_POST'
export const ADD_ATTACHMENT_RESPONSE = 'ADD_ATTACHMENT_RESPONSE'
export const REMOVE_ATTACHMENT = 'REMOVE_ATTACHMENT'
export const SIGN_IN_CHANGE = 'SIGN_IN_CHANGE'
export const SIGN_IN = 'SIGN_IN'
export const SIGN_IN_EMPTY = 'SIGN_IN_EMPTY'
export const SIGN_UP_CHANGE = 'SIGN_UP_CHANGE'
export const SIGN_UP = 'SIGN_UP'
export const SIGN_UP_EMPTY = 'SIGN_UP_EMPTY'
export const UPDATE_USER = 'UPDATE_USER'
export const USER_SETTINGS_CHANGE = 'USER_SETTINGS_CHANGE'
export const UPDATE_FRONTEND_PREFERENCES = 'UPDATE_FRONTEND_PREFERENCES'
export const UPDATE_PASSWORD = 'UPDATE_PASSWORD'
export const UPDATE_USER_PHOTO = 'UPDATE_USER_PHOTO'
export const GET_SINGLE_POST = 'GET_SINGLE_POST'
export const BAN = 'BAN'
export const UNBAN = 'UNBAN'
export const SUBSCRIBE = 'SUBSCRIBE'
export const UNSUBSCRIBE = 'UNSUBSCRIBE'
export const SEND_SUBSCRIPTION_REQUEST = 'SEND_SUBSCRIPTION_REQUEST'
export const GET_USER_COMMENTS = 'GET_USER_COMMENTS'
export const GET_USER_LIKES = 'GET_USER_LIKES'
export const EXPAND_SEND_TO = 'EXPAND_SEND_TO'
export const TOGGLE_HIDDEN_POSTS = 'TOGGLE_HIDDEN_POSTS'
export const SUBSCRIBERS = 'SUBSCRIBERS'
export const SUBSCRIPTIONS = 'SUBSCRIPTIONS'
export const GET_USER_INFO = 'GET_USER_INFO'
export const CREATE_GROUP = 'CREATE_GROUP'
export const UPDATE_GROUP = 'UPDATE_GROUP'
export const GROUP_REQUESTS = 'GROUP_REQUESTS'
export const ACCEPT_GROUP_REQUEST = 'ACCEPT_GROUP_REQUEST'
export const REJECT_GROUP_REQUEST = 'REJECT_GROUP_REQUEST'
export const ACCEPT_USER_REQUEST = 'ACCEPT_USER_REQUEST'
export const REJECT_USER_REQUEST = 'REJECT_USER_REQUEST'
