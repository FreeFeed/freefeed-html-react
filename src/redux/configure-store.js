import {compose, createStore, applyMiddleware, combineReducers} from 'redux';
import {createHistory, createHashHistory} from 'history';

import {
  apiMiddleware,
  authMiddleware,
  likesLogicMiddleware,
  userPhotoLogicMiddleware,
  groupPictureLogicMiddleware,
  redirectionMiddleware,
  requestsMiddleware,
  realtimeMiddleware,
  dataFixMiddleware
} from './middlewares';

import {routerReducer} from 'react-router-redux';
import * as reducers from './reducers';

//order matters â€” we need to stop unauthed async fetching before request, see authMiddleware
const middleware = [
  authMiddleware,
  apiMiddleware,
  dataFixMiddleware,
  likesLogicMiddleware,
  userPhotoLogicMiddleware,
  groupPictureLogicMiddleware,
  redirectionMiddleware,
  requestsMiddleware,
  realtimeMiddleware
];

const isDevelopment = process.env.NODE_ENV != 'production';

let enhancers = [applyMiddleware(...middleware),];

//tells webpack to include devtool enhancer in dev mode
if (isDevelopment && window.devToolsExtension) {
  enhancers.push(window.devToolsExtension());
}

const storeEnhancer = compose(...enhancers);

const createStoreWithMiddleware = storeEnhancer(createStore);
const reducer = combineReducers({...reducers, routing: routerReducer});

export default function configureStore(initialState) {
  return createStoreWithMiddleware(reducer, initialState);
}
