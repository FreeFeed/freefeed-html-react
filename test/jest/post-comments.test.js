/* global describe, it, expect, jest */
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom/extend-expect';
import { createStore } from 'redux';
import * as reactRedux from 'react-redux';

import PostComments from '../../src/components/post-comments';

// mock icon component to make snapshots smaller
jest.mock('../../src/components/fontawesome-icons', () => ({
  Icon: ({ icon }) => `fontawesome-icon ${icon.iconName}`,
}));

// mock it out because I don't want to deal with it
jest.mock('../../src/components/prevent-page-leaving', () => ({
  PreventPageLeaving: () => '',
}));

const AUTHOR = {
  id: 'user-id',
  username: 'author',
};

const VIEWER = {
  ...AUTHOR,
  frontendPreferences: {
    timeDisplay: {},
    comments: {
      highlightComments: true,
    },
    timeDifferenceForSpacer: 1000 * 60 * 60 * 24 * 6,
  },
};

const COMMENT1 = {
  id: 'comment1',
  body: 'Comment 1 body',
  omitBubble: true,
  user: AUTHOR,
  createdAt: '1617720000000', // Tue Apr 06 2021 14:40:00 GMT+0000
  seqNumber: 1,
};

const COMMENT2 = {
  id: 'comment2',
  body: '@author comment 2\nMultilinebody',
  omitBubble: true,
  user: { id: 'other-user', username: 'other' },
  createdAt: '1617780000000', // Wed Apr 07 2021 07:20:00 GMT+0000
  seqNumber: 2,
};

const COMMENT3 = {
  id: 'comment3',
  body: '^ response to comment 2',
  omitBubble: true,
  user: AUTHOR,
  createdAt: '1630000000000', // Thu Aug 26 2021 17:46:40 GMT+0000
  seqNumber: 3,
};

const POST = {
  id: 'post-id',
  omittedComments: 0,
  omittedCommentsOffset: 0,
  omittedCommentLikes: 0,
  isCommenting: true,
  isSinglePost: true,
  user: AUTHOR,
  createdAt: '1500000000000', // Fri Jul 14 2017 02:40:00 GMT+0000
};

const defaultState = {
  user: VIEWER,
  commentEditState: {},
  comments: {
    [COMMENT1.id]: COMMENT1,
    [COMMENT2.id]: COMMENT2,
    [COMMENT3.id]: COMMENT3,
  },
};

const renderPostComments = (props = {}, options = {}) => {
  const { Provider } = reactRedux;
  const dummyReducer = (state) => state;
  const store = createStore(dummyReducer, defaultState);

  const defaultProps = {
    user: VIEWER,
    post: POST,
    comments: [COMMENT1, COMMENT2, COMMENT3],
    entryUrl: 'https://',
    isSinglePost: true,
    isLoadingComments: false,
    addComment: () => {},
    toggleCommenting: () => {},
  };

  const rendered = render(
    <Provider store={store}>
      <PostComments {...defaultProps} {...props} />
    </Provider>,
    options,
  );

  return {
    ...rendered,
    rerender: (props = {}, options = {}) =>
      renderPostComments(props, { container: rendered.container, ...options }),
  };
};

describe('PostComments', () => {
  it("Renders post comments and doesn't blow up", () => {
    const { asFragment } = renderPostComments();
    expect(asFragment()).toMatchSnapshot();
  });

  it('Renders omitted comments and expands them', () => {
    const showMoreComments = jest.fn();
    const { rerender } = renderPostComments({
      post: {
        ...POST,
        omittedComments: 12,
        omittedCommentLikes: 34,
        omittedCommentsOffset: 1,
      },
      showMoreComments,
    });
    expect(screen.getByLabelText(/15 comments/)).toBeInTheDocument();
    userEvent.click(screen.getByText('12 more comments with 34 likes', { role: 'button' }));
    expect(showMoreComments).toHaveBeenCalledWith('post-id');

    expect(screen.queryByLabelText('Loading comments...')).not.toBeInTheDocument();
    rerender({
      post: {
        ...POST,
        omittedComments: 12,
        omittedCommentLikes: 34,
        omittedCommentsOffset: 1,
        isLoadingComments: true,
      },
      showMoreComments,
    });
    expect(screen.queryByLabelText('Loading comments...')).toBeInTheDocument();
  });

  it('Renders add comment link which toggles comment form', () => {
    const toggleCommenting = jest.fn();
    renderPostComments({
      user: VIEWER,
      post: {
        ...POST,
        isSinglePost: false,
        isCommenting: false,
      },
      toggleCommenting,
    });
    userEvent.click(screen.getByText('Add comment', { role: 'button' }));
    expect(toggleCommenting).toHaveBeenCalledWith('post-id');
  });

  it('Highlights a comment when arrow is hovered', () => {
    renderPostComments();
    expect(document.querySelectorAll('.highlighted').length).toBe(0);
    userEvent.hover(screen.getByText('^'));
    expect(document.querySelectorAll('.highlighted').length).toBe(1);
    expect(document.querySelector('.highlighted').getAttribute('data-author')).toBe('other');
  });
});
