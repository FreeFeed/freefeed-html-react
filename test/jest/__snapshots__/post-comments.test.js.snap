// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostComments Renders post comments and doesn't blow up 1`] = `
<DocumentFragment>
  <div
    aria-label="3 comments"
    class="comments"
    role="list"
  >
    <div
      aria-hidden="true"
      class="spacer"
    >
      <div
        class="icon"
      >
        fontawesome-icon clock
      </div>
      <div
        class="body"
      >
        over 3 years later
      </div>
    </div>
    <div
      class="comment my-comment"
      data-author="author"
      role="comment listitem"
    >
      <div
        class="comment-likes-container"
      >
        <div
          class="comment-likes"
        >
          <div
            aria-label="Like this comment"
            class="comment-heart"
            role="button"
          >
            fontawesome-icon heart
          </div>
        </div>
        <time
          class="comment-time"
          datetime="2021-04-06T14:40:00.000Z"
          title="Apr 6, 2021"
        >
          fontawesome-icon comment
        </time>
      </div>
      <div
        class="comment-body"
      >
        <div
          class="expandable expanded"
        >
          <span
            class="Linkify"
            dir="auto"
            role="region"
          >
            Comment 1 body
          </span>
          <span
            aria-label="Comment by author"
            class="comment-tail"
          >
             - 
            <span
              class="comment-tail__item"
            >
              <span
                class="user-name-wrapper"
              >
                <a
                  class=""
                >
                  <span
                    dir="ltr"
                  >
                    You
                  </span>
                </a>
              </span>
            </span>
             - 
            <span
              class="comment-tail__item comment-tail__actions"
            >
              <span
                class="comment-tail__action"
              >
                <a
                  aria-disabled="false"
                  aria-expanded="false"
                  aria-haspopup="true"
                  class="comment-tail__action-link comment-tail__action-link--more"
                  role="button"
                  tabindex="0"
                >
                  <span
                    class="root"
                  >
                    more
                  </span>
                </a>
              </span>
            </span>
          </span>
        </div>
      </div>
    </div>
    <div
      class="comment omit-bubble"
      data-author="other"
      role="comment listitem"
    >
      <div
        class="comment-likes-container"
      >
        <div
          class="comment-likes"
        >
          <div
            aria-label="Like this comment"
            class="comment-heart"
            role="button"
          >
            fontawesome-icon heart
          </div>
        </div>
        <time
          class="comment-time"
          datetime="2021-04-07T07:20:00.000Z"
          title="Apr 7, 2021"
        >
          <span
            class="comment-icon feed-comment-dot"
            id="comment-comment2"
          />
        </time>
      </div>
      <div
        class="comment-body"
      >
        <div
          class="expandable expanded"
        >
          <span
            class="Linkify"
            dir="auto"
            role="region"
          >
            <span>
              <span
                class="user-name-wrapper"
              >
                <a
                  class=""
                >
                  <span
                    dir="ltr"
                  >
                    @author
                  </span>
                </a>
              </span>
               comment 2
            </span>
            <br />
            <span>
              Multilinebody
            </span>
          </span>
          <span
            aria-label="Comment by other"
            class="comment-tail"
          >
             - 
            <span
              class="comment-tail__item"
            >
              <span
                class="user-name-wrapper"
              >
                <a
                  class=""
                >
                  <span
                    dir="ltr"
                  >
                    other
                  </span>
                </a>
              </span>
            </span>
             - 
            <span
              class="comment-tail__item comment-tail__actions"
            >
              <span
                class="comment-tail__action"
              >
                <a
                  aria-disabled="false"
                  aria-expanded="false"
                  aria-haspopup="true"
                  class="comment-tail__action-link comment-tail__action-link--more"
                  role="button"
                  tabindex="0"
                >
                  <span
                    class="root"
                  >
                    more
                  </span>
                </a>
              </span>
            </span>
          </span>
        </div>
      </div>
    </div>
    <div
      aria-hidden="true"
      class="spacer"
    >
      <div
        class="icon"
      >
        fontawesome-icon clock
      </div>
      <div
        class="body"
      >
        5 months later
      </div>
    </div>
    <div
      class="comment omit-bubble my-comment"
      data-author="author"
      role="comment listitem"
    >
      <div
        class="comment-likes-container"
      >
        <div
          class="comment-likes"
        >
          <div
            aria-label="Like this comment"
            class="comment-heart"
            role="button"
          >
            fontawesome-icon heart
          </div>
        </div>
        <time
          class="comment-time"
          datetime="2021-08-26T17:46:40.000Z"
          title="Aug 26, 2021"
        >
          <span
            class="comment-icon feed-comment-dot"
            id="comment-comment3"
          />
        </time>
      </div>
      <div
        class="comment-body"
      >
        <div
          class="expandable expanded"
        >
          <span
            class="Linkify"
            dir="auto"
            role="region"
          >
            <span
              class="arrow-span"
            >
              ^
            </span>
             response to comment 2
          </span>
          <span
            aria-label="Comment by author"
            class="comment-tail"
          >
             - 
            <span
              class="comment-tail__item"
            >
              <span
                class="user-name-wrapper"
              >
                <a
                  class=""
                >
                  <span
                    dir="ltr"
                  >
                    You
                  </span>
                </a>
              </span>
            </span>
             - 
            <span
              class="comment-tail__item comment-tail__actions"
            >
              <span
                class="comment-tail__action"
              >
                <a
                  aria-disabled="false"
                  aria-expanded="false"
                  aria-haspopup="true"
                  class="comment-tail__action-link comment-tail__action-link--more"
                  role="button"
                  tabindex="0"
                >
                  <span
                    class="root"
                  >
                    more
                  </span>
                </a>
              </span>
            </span>
          </span>
        </div>
      </div>
    </div>
    <div
      aria-hidden="true"
      class="spacer"
    >
      <div
        class="icon"
      >
        fontawesome-icon clock
      </div>
      <div
        class="body"
      >
        about 4 years later
      </div>
    </div>
    <div
      class="comment"
      data-author=""
      role="comment listitem"
    >
      <div
        class="comment-likes-container"
      >
        <span
          class="comment-time"
        >
          fontawesome-icon comment
        </span>
      </div>
      <div
        class="comment-body"
        role="form"
      >
        <div>
          <textarea
            class="comment-textarea"
            dir="auto"
            maxlength="3000"
          />
        </div>
        <div>
          <button
            aria-disabled="true"
            aria-label="Submit disabled (textarea is empty)"
            class="btn btn-default btn-xs comment-post disabled"
          >
            Comment
          </button>
          <span
            class="submit-mode-hint"
          >
            (Enter for new line)
          </span>
          <a
            aria-disabled="false"
            class="comment-file-button iconic-button"
            role="button"
            tabindex="0"
            title="Add photo or file"
          >
            fontawesome-icon paperclip
          </a>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
